src/unix/fs.c

```diff
@@ -1055,18 +1055,14 @@ static ssize_t uv__fs_copyfile(uv_fs_t* req) {
 #ifdef FICLONE
   if (req->flags & UV_FS_COPYFILE_FICLONE ||
       req->flags & UV_FS_COPYFILE_FICLONE_FORCE) {
-    if (ioctl(dstfd, FICLONE, srcfd) == -1) {
-      /* If an error occurred that the sendfile fallback also won't handle, or
-         this is a force clone then exit. Otherwise, fall through to try using
-         sendfile(). */
-      if (errno != ENOTTY && errno != EOPNOTSUPP && errno != EXDEV) {
-        err = UV__ERR(errno);
-        goto out;
-      } else if (req->flags & UV_FS_COPYFILE_FICLONE_FORCE) {
-        err = UV_ENOTSUP;
-        goto out;
-      }
-    } else {
+    if (ioctl(dstfd, FICLONE, srcfd) == 0) {
+      /* ioctl() with FICLONE succeeded. */
+      goto out;
+    }
+    /* If an error occurred and force was set, return the error to the caller;
+     * fall back to sendfile() when force was not set. */
+    if (req->flags & UV_FS_COPYFILE_FICLONE_FORCE) {
+      err = UV__ERR(errno);
```

Comment:
I'm not sure how others feel, but I have a slight preference toward keeping `err` as `UV_ENOTSUP` in this case.

Reply:
Agreed.

Reply:
Thanks for reviewing

> I'm not sure how others feel, but I have a slight preference toward keeping `err` as `UV_ENOTSUP` in this case.

I thought about that as it would change the API. But in my opinion it is not a good idea to mess around with the return values, especially of system calls. Here it obfuscates the real reason that `ioctl()` failed (see `man 2 ioctl`), making it possibly unnecessarily difficult to debug.

However, I agree that this should be separate from the fix. I'll prepare a clean PR for the fix alone and propose the other change separately.

Reply:
@cjihrig @saghul I can see cases where this is useful. If FICLONE isn't supported, errno is probably EINVAL but if e.g. `srcfd` got closed concurrently then it'll return EBADF.

Reply:
That's a great point! 👍  to leaving it as-is.

Reply:
Yea, that SGTM

Reply:
> @cjihrig @saghul I can see cases where this is useful. If FICLONE isn't supported, errno is probably EINVAL but if e.g. `srcfd` got closed concurrently then it'll return EBADF.

Thanks for the explanation!
Indeed one of the yarn issues is about copyfile throwing `EBADF`, the other one gets `EFAULT`.